### DISCUSSION TITLE
Inverse split tunneling

### DISCUSSION THREAD

**Discussion by nalllen** [mullvad/mullvadvpn-app#9050] [Feature requests & Ideas]
The recent beta release that added split tunneling to windows is great, but its sadly not perfect for my use case since trying to add almost everything to the splitt tunneling list basically becomes unsustainable.

so i would really like a inverse split tunneling feature that only routes selected programs or ip's and leave rest unaffected.

a use case for this could be when performance and latency is very important for the majority of programs and games you run.

This might go against mullvads philosophy regarding privacy, but i feel it would be a great option to have!

**Comment by pinkisemils** [comment]
You should be able to exclude Steam and have all games launched by Steam to be excluded. Same goes for specific shells. But we'll consider this feature too.

**Comment by catwithbanana** [comment]
I would also like to see this feature added. The Mullvad app doesn't pick up a lot of my app installs, and hunting down every game and professional program I need to exclude is a gargantuan task.

It would be a big boost to usability if I could choose to only tunnel a few crucial programs, like web browsers, while leaving everything else untouched.

**Comment by Inverness** [comment]
I would also like to voice my support for this if it will help.

I want the best performance and least latency for the majority of programs I use. There are only a handful that I'm concerned with securing with a VPN.

I'd suggest an option for each program in the split tunneling list as to whether or not it will travel through the VPN, and then an option for whether all programs that are not in the list will go through the VPN or not.

This would offer a good level of control for all kinds of users.

**Comment by faern** [comment]
We will likely not implement this on Windows at least. Because we have looked at what would be needed and it's more work than a simple negation of some firewall rules. All these rules are pretty critical to the security of the apps and any extra logic introduces new risks of having bugs causing leaks that could be critical. We don't feel like the current need for this is large enough to justify such a risk. After all, the app is privacy oriented and the intended use case is that it tunnels all your traffic.

Since we currently can't make DNS on Windows go outside the tunnel for an excluded process it would be strange to exclude most processes and only tunnel a few. Because the DNS requests for all excluded processes would still go in the tunnel. And that's likely not what you would expect/want as a user only tunneling a few applications. All the applications outside the tunnel will get DNS responses as if they were in the tunnel, which could affect their behavior and functionality.

**Reply by asdffdsazqqq** [reply]
what a bunch of nonsense. i have posted exactly how to that on windows, and it is trivial. 
and the fact is, other vpn providers have done that for years.

i believe the real reason is that mullvad vpn is stagnant, having financial issues, cannot afford to add a feature that is simple to implement and most wanted by paying customers.

for example, for two weeks now, at https://mullvad.net/en/servers.
"We are currently experiencing issues with our DAITA servers, we are looking into this"

why two weeks and still no fix?

**Comment by Johnyknowhow** [comment]
That's a shame. Having a whitelist would be most useful as I only really want to use Mullvad for IP masking for my torrent clients. I tried setting up OpenVPN with Mullvad using _route-nopull_ and trying to get only my torrent client to tunnel through, but I just couldn't get it to operate properly, through any combination of port forwarding with MV, running the SOCKS5 proxy or not, and changing my settings, some trackers just wouldn't connect, UDP seemed to be all jacked, I got overwhelmed and gave up. Tunneling through the Mullvad App seemed to work fine though (and was much simpler; i admit to my lack of experience), so I opted to just use it and whitelist every other program I could think of, although I know there's lots still that aren't in the list that I'd have to add and manually browse for their executables. 

Even though you can't implement a way for excluded applications to route DNS information outside of the tunnel, I don't really mind if all of my DNS traffic goes through the VPN tunnel anyway, as I only am really using Mullvad servers in my own geographical location, since I only want my IP to be masked, I don't really want to show up as a different country. 

It would be a nice-to-have, but if it's far too much trouble to implement, I suppose the wants of a select handful of us users aren't that high on the development docket haha

**Comment by jonpolak** [comment]
Will *inverse split tunneling* be implemented for LINUX? 

I've tried to implement this with the [namespace technique](https://www.wireguard.com/netns/)  but have not managed to get it to work with Mullvad. If there is a way to do it manually, with the namespace trick, then that would be good enough for me, but would love to have formal validated instructions from Mullvad on how to do it properly ensuring that DNS goes to the right tunnel/gateway and everything stays nicely segregated. 

I was able to do this (with openvpn) before _systemd-resolved_ took over DNS in Ubuntu. Now it's quite difficult to figure out what that resolver is doing and I rely on dnsleaktest.com to tell me!

**Comment by cooky-cook** [comment]
This is a real shame since certain competitors offer much more flexible settings - whitelist/blacklist and separate DNS for direct/VPN connections. The (apparent) superiority and configuration flexibility is one of the reasons why I haven't switched to Mullvad yet.

![2021-11-05_113303](https://user-images.githubusercontent.com/57732482/140498437-44160474-d937-47fd-b21b-de4b07456263.png)

Their VPN client is open source. Why cannot Mullvad offer such a flexible configuration - is it due to security (if their implementation is inherently insecure, I think this fact should be made known) or not enough resources allocated/not seeing this as an important feature?

**Comment by dnet890** [comment]
@cooky-cook That PIA tunneling feature seems quite overwhelming and I know Mullvad is good for simplicity and less cluttered. So, I hope they can make it more simple.

**Comment by mikamidd** [comment]
This is something I'd love to see implemented for Windows too eventually.
In addition to PIA, I believe ProtonVPN also has both normal and inverse split tunnel, their apps are open source too.

**Comment by Apposite245** [comment]
Chiming in for support of this feature. I really only use Mullvad for one or two programs so having to manually split tunnel everything as I install new programs is becoming a headache.

**Comment by UnknownProgrammer-ttu** [comment]
Adding another comment to the pool of people who want this feature.

**Comment by VoidCruzer** [comment]
Adding another for someone who would like to see this implemented. This is the biggest feature I miss from ProtonVPN that Mullvad doesn't have. The way it is setup now is quite an annoyance to go through and find every program and is almost impossible for some games and anti-cheat etc.

**Comment by trunkensailor** [comment]
For a linux implementation, [vopono](https://github.com/jamesmcm/vopono) would be a good reference implementation as doing split tunneling this way can be very useful.

I for one would really like to see this on android, though. [OpenVPN](https://github.com/schwabe/ics-openvpn) already supports this (you can switch between a white- and a blacklist for apps that should be connected to the VPN) and I don't think this would be hard to implement.

**Comment by dedors** [comment]
I like the mullvad app, but this feature missing just made me go back to the horrible openvpn solution.

**Comment by unkn4wn** [comment]
This feature is absolutely necessary because many origin games don't work while split tunneling even if you exclude the origin launcher + exe files manually. Maybe I am missing an important exe file which I need to exclude too but if we had an "inversed split tunnel" the problem would be solved easily.

**Comment by naodesu** [comment]
Because official app for Windows and Android lacks this feature I used official WireGuard app and SOCKS proxies.
This is possible, but hard to configure, and needs regular maintenance (because mullvad server changes). Also, Mullvad SOCKS proxies are slow compared to pure wireguard protocol. 
So in the end I switched to app, but need to exclude tons of programs, because I really need Mullvad only for 1-2 apps.

**Comment by GrimWreeper** [comment]
I would _really like_ this feature. Split tunnelling in it's current implementation is backwards in my opinion. I'd find it more useful and feel more secure with a small list of apps that go through my Wireguard tunnel, knowing everything else is normal ISP traffic.
The current implementation is _so close_ to fitting my use case, just needs to be inverted.

**Comment by sendarion** [comment]
Maybe this can help you guys meanwhile : https://asheroto.medium.com/split-tunneling-in-wireguard-on-windows-e2dfd86d5982

**Comment by Pohra** [comment]
I would like this feature as well, primarily because it seems impossible to actually add Windows Store apps/Xbox games to split tunneling because of the hyper-locked-down file security. I can't seem to modify permissions to allow Mullvad to see those programs without breaking them. Inverse split tunneling would work around that problem by letting me opt-in instead of opt-out.

Many other products allow you choose between opt-in and opt-out, so I know it can be done.

**Comment by DAOWAce** [comment]
Another voice for support of this feature.  Would help UWP apps as well: https://github.com/mullvad/mullvadvpn-app/issues/2822

My usage of a VPN most definitely is the opposite of the default; only want it routing torrent clients.. but instead have to blacklist every single other damn program I use and game I play, which is incredibly annoying.  Ontop of that, certain things (some [game connections](https://i.imgur.com/AiFgUvm.png)) just don't seem to work properly when added to the tunnel and I can't figure out why.

Also, outside of the scope of this issue, but the default list of programs is.. very strange.  It seems like it's only pulling a list from the start menu programs, and there's no way to change it.  There's a lot of irrelevant things in there, especially if you have a lot of GOG games installed:  
![Mullvad_VPN_2022-12-01_13-31-18](https://user-images.githubusercontent.com/480587/205132300-8ad29ade-b06a-4af1-8a44-82e319e6ffd1.png)

And lastly, when manually choosing an exe, it always defaults to the user folder instead of the last selected location.  Minor gripe in comparison.

**Comment by EricIsbell** [comment]
i also agree. An inverse split tunnelling feature would be an absolute game changer. Being able to choose what applications you want to be in the vpn would be a life saver. It's just too hard and takes way too much time and effort to add every single program i dont want to be connected to the vpn.

**Comment by ganeshh123** [comment]
I would also like this to be implemented. 

Currently I am using [wiresocks](https://github.com/ivoronin/wiresocks) as a workaround, with wireguard config files downloaded from https://mullvad.net/account . It's a decent solution for applications that support socks5 proxies.

**Comment by agent-seed** [comment]
Just adding my voice to the pool. I can't seem to get minecraft excluded no matter what exes I select.

**Comment by ganeshh123** [comment]
> Just adding my voice to the pool. I can't seem to get minecraft excluded no matter what exes I select.

You can try using something like NetLimiter to monitor which processes Minecraft uses to send and receive data

**Comment by onionrepo** [comment]
Hi, I would also love to see this added to the Windows build. When I'm trying to only run my browser through a VPN it's a pain to add every application and then manually remove them all when I want to run everything through a VPN again.

**Comment by bredmor** [comment]
Commenting to add my voice to this request as well. My primary use-case for the VPN is getting alternative routing for specific applications when there's a problematic node between me and the server I'm connected to. I rarely want more than a single application to use the VPN.

**Comment by Sad-theFaceless** [comment]
Yes please, that would be super useful !

**Comment by crypaus** [comment]
FOR ANYONE THAT IS TRYING TO GET INCLUSIVE IS EASY, DOWNLOAD THE OPENVPN AND WIREGAURD CONFIG FROM THE SITE, AND THEN DOWNLOAD Windscribe_2.6.14.exe AND AFTER THAT JUST CREATE A FREE ACCOUNT AND UPLOAD YOUR CONFIG FILE TO THEIR PLATFORM, THEN CONNECT, AFTER THAT GO IN SETTINGS AND THEN TURN ON THE SPLIT. AFTER THAT YOU CAN CHOOSE IF YOU WANT IT INCLUSIVE OR EXCLUSIVE,

GOOD LUCK

**Comment by asdffdsazqqq** [comment]
hello, thanks, sorry if i lack the technique skills to follow this topic.

for years, this worked great -> mullvad over openvpn and this helpful guide, https://mullvad.net/en/help/split-tunneling-mullvad-vpn. just need access to the socks5 proxy servers, to use with firefox and mullvad browser with mullvad browser extension.
basically, all that is needed is `route-nopull` in the .ovpn config file.

do not want vpn to change the default route, do not want vpn to change the routing table, do not want all traffic to flow over vpn.
i spent days, tried `Table = off` and various attempts at `AllowedIPs`

just want to use socks5 proxy servers without forcing all traffic thru vpn. as i upload a lot of large files to internet, no reason or want to force that thru slow vpn connections


please, help, thanks so much,
david

**Comment by asdffdsazqqq** [comment]
> Maybe this can help you guys meanwhile : https://asheroto.medium.com/split-tunneling-in-wireguard-on-windows-e2dfd86d5982

i had already found that, tried that and while it did add/remove rules from routing table, still never worked for me.

the only problem with using mullvad over openvpn , it cannot access multiple proxy servers as compared to mullvad over wireguard.

sorry, a bit frustrating, after all this time, wireguard over mullvad still cannot do what competition can or even what mullvad over openvpn can?

**Comment by Qtax** [comment]
> do not want vpn to change the default route, do not want vpn to change the routing table, do not want all traffic to flow over vpn. i spent days, tried `Table = off` and various attempts at `AllowedIPs`

@asdffdsazqqq this worked for me on windows 10: https://superuser.com/a/1658611/89979

In summary:

1. Install WireGuard for Windows
2. Set `HKEY_LOCAL_MACHINE\Software\WireGuard\DangerousScriptExecution` registry to `1` by for example running this command in an admin shell: `reg add HKLM\Software\WireGuard /v DangerousScriptExecution /t REG_DWORD /d 1 /f`
3. Use the following WireGuard config:

```
[Interface]
PrivateKey = <...>
Address = <...>
DNS = <...>
PostUp = powershell -command "$wgInterface = Get-NetAdapter -Name %WIREGUARD_TUNNEL_NAME%; route add 0.0.0.0 mask 0.0.0.0 0.0.0.0 IF $wgInterface.ifIndex metric 95"
PreDown = powershell -command "$wgInterface = Get-NetAdapter -Name %WIREGUARD_TUNNEL_NAME%; route delete 0.0.0.0 mask 0.0.0.0 0.0.0.0 if $wgInterface.ifIndex metric 95"
Table = off

[Peer]
PublicKey = <...>
AllowedIPs = 0.0.0.0/0,::0/0
Endpoint = <...>
```

**Comment by asdffdsazqqq** [comment]
@Qtax, thanks, but that approach never fully worked for me, much as tried.
https://superuser.com/questions/1658610/how-do-i-disable-routing-table-changes-in-wireguard-for-windows/1658611#1658611 
is basically the same as this, as mentioned earlier in this topic.
https://asheroto.medium.com/split-tunneling-in-wireguard-on-windows-e2dfd86d5982

just now, after days for experimenting, i have a working setup. seems to do everything i need
* at the same time, access to openvpn socks5 and EVERY wireguard socks5
* works with mullvad browser extension, for easy switching of wireguard proxy servers
* works with firefox with multi-container extension, where each container uses a different proxy server

```
[Interface]
PrivateKey = redacted
Address = 10.68.138.201/32
PostUp = cmd /c route add 10.124.0.0 mask 255.255.254.0 10.68.138.201 metric 9999
PreDown = cmd /c route delete 10.124.0.0 mask 255.255.254.0
Table = off

[Peer]
PublicKey = redacted
AllowedIPs = 0.0.0.0/1, 128.0.0.0/1
Endpoint = 198.44.128.194:51820
```


```
curl https://ipv4.am.i.mullvad.net --max-time 2 --silent 
74.102.225.xxx

curl https://ipv4.am.i.mullvad.net --socks5-hostname  10.8.0.1 --max-time 2 --silent 
146.70.166.3

curl https://ipv4.am.i.mullvad.net --socks5-hostname 10.124.1.38 --max-time 2 --silent 
69.4.234.120

curl https://ipv4.am.i.mullvad.net --socks5-hostname us-slc-wg-socks5-103.relays.mullvad.net --max-time 2 --silent 
69.4.234.120
```

![image](https://github.com/mullvad/mullvadvpn-app/assets/90116442/31fbb064-79a1-491b-b739-ddfde550587e)

![image](https://github.com/mullvad/mullvadvpn-app/assets/90116442/27ab3040-ce98-4e0a-960d-531020f3a081)

**Comment by hianick** [comment]
Would be nice to see this sometime.

**Comment by tmcelroy2202** [comment]
Would also love this.

**Comment by chexo3** [comment]
Here to say that this would save me a lot of headaches. In fact, I'm cancelling my subscription right now and seeking a refund because this is a real dealbreaker. I can set it up manually with OpenVPN, sure, but like others have said that jacks things up. Like, the SOCKS proxy y'all provide doesn't seem to support IPv6 at all and UDP is jacked up so I can't use it for what I want to use it for

**Comment by JW-Rami** [comment]
> We will likely not implement this on Windows at least. Because we have looked at what would be needed and it's more work than a simple negation of some firewall rules. All these rules are pretty critical to the security of the apps and any extra logic introduces new risks of having bugs causing leaks that could be critical. We don't feel like the current need for this is large enough to justify such a risk. After all, the app is privacy oriented and the intended use case is that it tunnels all your traffic.
> 
> Since we currently can't make DNS on Windows go outside the tunnel for an excluded process it would be strange to exclude most processes and only tunnel a few. Because the DNS requests for all excluded processes would still go in the tunnel. And that's likely not what you would expect/want as a user only tunneling a few applications. All the applications outside the tunnel will get DNS responses as if they were in the tunnel, which could affect their behavior and functionality.

------------------
LISTEN TO YOUR CUSTOMER PLEASE !!!!! WE WANT IT SO MUCH !!!
@faern

**Comment by GuruOz** [comment]
Sad to see this feature is still not available. I would argue that a large portion of the customer base is looking to route only one particular application's traffic (We all know what this is).

I understand that there are difficulties implementing this on Windows, but I do not see any explanation for Linux. If this was available in Linux, I could at least come up with a viable workaround. 

Unfortunately, I was not able to figure out an easy way to set this up, so I have to cancel my subscription until this feature gets implemented.

**Comment by kaymanov-emitter** [comment]
Throwing my hat in as another that would love this feature.

**Comment by misurbraford** [comment]
Another call for this feature. 

Basically, I'm just trying to use my PC without mullvad VPN, and then only have a secure browser use mullvad (for example, mullvad browser).

**Comment by colemar** [comment]
At least on the Android app it would suffice to have a button that moves all apps at once to Excluded Apps, and another button that removes all apps at once from Excluded Apps.
Then I would be able to select the few apps that should connect via VPN without having to tap on about one hundred (+) buttons.

**Comment by colemar** [comment]
> This is a real shame since certain competitors offer much more flexible settings - whitelist/blacklist and separate DNS for direct/VPN connections. The (apparent) superiority and configuration flexibility is one of the reasons why I haven't switched to Mullvad yet.

But not for Android apparently. I just installed PIA app: it defaults on all app "protected" and you have to tap on each single one to leave only the few apps you want "protected".

**Comment by candroid-man** [comment]
IVPN has this feature, and I'm waiting for Mullvad to have this feature as well before I switch.

**Comment by Voleno1** [comment]
Add. This.

**Comment by GRIFFiin** [comment]
PLEASE add this 😭 how is it not in yet

**Comment by iamstevedavis** [comment]
+1

**Comment by OrkoGrayskull** [comment]
Please. Add. This. Feature. Mullvad!

**Comment by MediocreTwo-7** [comment]
Please add this, i cant play minecraft :( 
already tried everything but only shutting mullvad down seems to help.

**Comment by DAOWAce** [comment]
> Please add this, i cant play minecraft :( already tried everything but only shutting mullvad down seems to help.

Yeah, like I [mentioned in Dec 2022](https://github.com/mullvad/mullvadvpn-app/issues/2808#issuecomment-1334192801), it just seems impossible.  I've not found a way to do it either.

Sad this has been left stale for so long.

**Comment by voidblack** [comment]
+1

**Comment by Digitalone1** [comment]
> For a linux implementation, [vopono](https://github.com/jamesmcm/vopono) would be a good reference implementation as doing split tunneling this way can be very useful.

It would be very useful, but vopono is not a good reference implementation. It does not work with IPv6 endpoints (https://github.com/jamesmcm/vopono/issues/181) and even with IPv4 endpoints there are various issues when you have a running firewall (https://github.com/jamesmcm/vopono/issues/273). It never worked to me.

If namespaces are the only way of doing the inverse split tunnelling on Linux, they are too complex and will lead to other troubles that it's not worth to implement them.

**Comment by SeaOfMan** [comment]
bump since this is a much needed future, Especially for my use case.

**Comment by janiaul** [comment]
Please add this feature.

I only want to use VPN for few apps on my Windows PC and having to manually exclude every exe is very overwhelming and requires lots of maintenance.

**Comment by meotimdihia** [comment]
yeah, we often only want to use VPN for browsers or some apps, not everything. 

The default should be reverse current split tunnel.

**Comment by numfin** [comment]
cant work with mullvad app, We using wireguard at work. Work subnet not accessible when mullvad active. Found this issue - please implement, or dont add mullvad routes at metric 1

**Comment by rentoggle** [comment]
this could be very useful since i want to tunnel only some apps and not go through every single application enabling split-tunneling on them

please add this feature

**Comment by biggestsonicfan** [comment]
+1

EDIT: I've had to wireguard within a docker to get the results I wanted 😔

**Comment by lazymonkey2** [comment]
I know you don't care but I moved to proton vpn, which has this feature.

**Comment by pierrepgithub** [comment]
bump +1

**Comment by dangoquidev** [comment]
+1, really want this feature, your vpn is kinda the best one I can use in China and this feature can be game changer for me

**Comment by colemar** [comment]
I would like to keep using Mullvad, but this missing feature is too important, therefore I switched to another vpn provider.
Even the plain Wireguard Android client has an app filter much more manageable, and I have a remote host in France that I can use as vpn server.

**Comment by dedors** [comment]
Still a highly wanted feature

**Comment by pacopepepipo** [comment]
+1

i still can't understand why it takes years to add this simple option

**Comment by kyngs** [comment]
Hey,
still no update on this? This is probably the most wanted issue ATM, and ignoring it won't make it go away.
I understand that due to technical reasons, it's not feasible on Windows (even though competitors manage to do so), however, can we at least get some basic implementation for Linux (or at least a timeframe for it)?

**Comment by pacopepepipo** [comment]
> Hey, still no update on this? This is probably the most wanted issue ATM, and ignoring it won't make it go away. I understand that due to technical reasons, it's not feasible on Windows (even though competitors manage to do so), however, can we at least get some basic implementation for Linux (or at least a timeframe for it)?

how is not possible on windows when there's not only competition to do it, but powershell commands to do it?

even vpn clients like wiresock or wireguard do it creating your own vpn's to vps.

i mean, they don't want to do it, ok, then just say "deal with it, it's too much work and we don't want to dedicate resources to it", but feasible it is.

**Comment by drdzyk** [comment]
Hi, please get it finally done!

**Comment by JaxxArmstrong** [comment]
Bump +1

This would make my life so much easier, thank you.

**Comment by HBRJZ** [comment]
Another +1, please add this feature to the app for Windows/Linux/MacOS/iOS.

**Comment by virtualgurel** [comment]
+1

**Comment by devingDev** [comment]
Wow 2021... just literally invert the logic of blacklist? how difficult can this be and still not fixed. Yes not FIXED. This is a bug imho or at least a crucial missing feature. I need to only use vpn for a single app since its blocked here and the rest i can use fine. Please show some effort, it's been almost 4  years

**Comment by FeyrisTan** [comment]
It's been almost 4 years... Damn.
Really does not seem like they care.

**Comment by LasaJoniHD** [comment]
+1

**Comment by Digitalone1** [comment]
> Wow 2021... just literally invert the logic of blacklist? how difficult can this be and still not fixed. Yes not FIXED. This is a bug imho or at least a crucial missing feature. I need to only use vpn for a single app since its blocked here and the rest i can use fine. Please show some effort, it's been almost 4 years

Don't know for other platforms, but on Linux it's almost impossible. The only way is to make it with namespaces like vopono, but it does not work for IPv6 endpoints and, even on IPv4 ones, there are issues when a firewall is installed. See https://github.com/jamesmcm/vopono/issues/273

TLDR: It's not worth to implement this feature.

**Comment by dedors** [comment]
I did switch to another provider only because of this. Bye and everyone good luck with getting it someday. I loved mullvad for everything else.

**Comment by djodars89** [comment]
+1
I only want to VPN a single browser and not all my OS / apps.
It would therefore be infinitely easier to add the desired executable(s) in the list instead of excluding dozens or even hundreds of apps.

**Comment by asdffdsazqqq** [comment]
> I only want to VPN a single browser and not all my OS / apps.

hi, i have been doing that for many years.
in my case, i want just firefox and mullvad browser to use the vpn but works with any app that can access a socks5 proxy server.
for instance, i have used this techinque with rclone and curl.

1. instead of mullvad app, use any openvpn app. i use viscosity.
2. in firefox+mullvad browser, set "SOCKS Host" to `10.8.0.1`
3. using mullvad website, export a .opvn with as many endpoints as you want. i choose every single endpoint in USA and use `remote-random` in the .conf file.
4. make sure do not set a default route and default gateway

here are my settings from viscosity

![Image](https://github.com/user-attachments/assets/7d338019-1332-4dab-aea0-276924dc20bd)

![Image](https://github.com/user-attachments/assets/e7a81e5a-ce91-4c63-8b3f-40d8307e6596)

![Image](https://github.com/user-attachments/assets/2ea8351b-0f8a-48d2-a3ff-0c490a5e6a2e)

```
remote-random
pull-filter ignore route-ipv6
pull-filter ignore ifconfig-ipv6
pull-filter ignore "redirect-gateway ipv6"
route-nopull
resolv-retry infinite
verb 3
sndbuf 524288
rcvbuf 524288
cipher AES-256-GCM
tls-cipher TLS-DHE-RSA-WITH-AES-256-GCM-SHA384
service mullvadopenvpn
block-outside-dns
dev-node {3DA44DB8-B007-4A47-AA10-091726F18616}
```

**Comment by ynyxchen** [comment]
+1 for this

**Comment by BugsWorld** [comment]
Please add this option. Leave it up to the user to choose if they want full security or not.

This is very needed, especially today where many services are banning VPN IPs. But that's just one of the reasons why this is needed. Please consider it, as a paying customer, I would say this is mandatory.

**Comment by curious-youna** [comment]
+1 You will get tons of customers more if you added this

**Comment by Vansinnet** [comment]
Any chance we will get this soon? If Mullvad is concerned about security, why just not implement this and if the user wants to do inverted split tunneling, warn the user before it can be enabled.

**Comment by ApexProprietary** [comment]
The UK has now made it law that all sites that _can_ show adult content require you to send them your ID or banking details, making this feature the only potential solution to making the internet actually usable here. Having to whitelist every single program on my PC except for my browser and discord is simply not viable.

**Comment by VioletKitten** [comment]
changing provider because of this. lame.

**Comment by Retribution1337** [comment]
> The UK has now made it law that all sites that _can_ show adult content require you to send them your ID or banking details, making this feature the only potential solution to making the internet actually usable here. Having to whitelist every single program on my PC except for my browser and discord is simply not viable.

This is pretty much exactly what I'm using this for on my machine. Primarily used it on my media server for downloads but now that this is happening, spooled it up on my desktop to make sure firefox and discord go through it.

Inverse Split Tunneling would be great.

**Comment by Lifell** [comment]
I would also very much like this feature. Just dropping my hand on the pile I suppose.

> We will likely not implement this on Windows at least. Because we have looked at what would be needed and it's more work than a simple negation of some firewall rules. All these rules are pretty critical to the security of the apps and any extra logic introduces new risks of having bugs causing leaks that could be critical. We don't feel like the current need for this is large enough to justify such a risk. After all, the app is privacy oriented and the intended use case is that it tunnels all your traffic.
> 
> Since we currently can't make DNS on Windows go outside the tunnel for an excluded process it would be strange to exclude most processes and only tunnel a few. Because the DNS requests for all excluded processes would still go in the tunnel. And that's likely not what you would expect/want as a user only tunneling a few applications. All the applications outside the tunnel will get DNS responses as if they were in the tunnel, which could affect their behavior and functionality.

Is this still the case considering it has been 4 years?

**Comment by LuminousToucan** [comment]
FYI if you're having issues with Steam, closing and relaunching Steam _after_ toggling on Mullvad (with Steam excluded in split-tunnel config) makes it not use the VPN connection and unblocks downloads.

Parsec I've had 0 luck with.

+1 for the inverse split-tunnel idea. I imagine demand for a feature like this has spiked massively as of late. Can't speak to the technical limitations or implementation and it's a feature which inherently adds more insecurity, but the market of normies who want to only route Discord and a browser through a VPN shot up by about 50 million recently, and things don't look like they're getting better soon - just a thought.

**Comment by MulvadViking** [comment]
Backing this feature—Mullvad’s only needed for some apps, so not looking to manually setting up split.

**Comment by DAOWAce** [comment]
> I see why the devs may be holding back

I think it's simply due to a philosophical reason instead of a technical one.

There is no reason why most other VPNs have had inverse split tunneling for over a decade, yet one of the highest rated VPNs that advertises itself as the most "secure and private" lacks an option that goes against their design philosophy.

Split tunneling was added, after all.  Inverting it doesn't seem like any challenge.  _(not a dev)_

I will repeat again for visibility: Certain programs (eg minecraft java, Microsoft UWP apps..) are completely incompatible with split tunneling for reasons unknown (to me).  Only an inverse split tunnel (or all VPN traffic, which is bad for server hosting) will let them run correctly.

**Comment by Hasshu** [comment]
What does the current approach mean for things like VoLTE and Wi-Fi calling?

**Comment by NeurekaSoftware** [comment]
I'm a new Mullvad user, and this is the one thing that bothers me. I may switch providers because of this not being an option.

**Comment by JG606** [comment]
Chiming in to also say that this is disappointing. I've been very happy with Mullvad up until now, but there are certain applications such as Java applications that I need to be able to exclude from the VPN, and cannot. I would also be satisfied with an inverse split tunnelling mode that allowed me to select apps to include rather than exclude. As it stands, I'm forced to switch VPN provider which is sad.

**Comment by vjt4** [comment]
Yeah I just refunded because this program does not have this feature.

**Comment by asdffdsazqqq** [comment]
what a bunch of nonsense. i have posted exactly how to that on windows, and it is trivial.
and the fact is, other vpn providers have done that for years.

i believe the real reason is that mullvad vpn is stagnant, having financial issues, cannot afford to add a feature that is simple to implement and most wanted by paying customers.

for example, for two weeks now, at https://mullvad.net/en/servers.
"We are currently experiencing issues with our DAITA servers, we are looking into this"

why two weeks and still no fix?

**Comment by faern** [comment]
Hi everyone participating in this discussion. I'm sadly going to close this as not planned. But I will with this post also explain why we are not going to implement inverse split tunneling, so you understand the reasoning.

TLDR:
* We think it's a dangerous foot-gun for our customers, that can hurt their privacy and security.
* On a technical level this would have huge implications for the app.

To quote the original discussion starter:

> This might go against mullvads philosophy regarding privacy, but i feel it would be a great option to have!

We understand that this feature would be really awesome for some! But yes, ultimately we have concluded that it does not align with our privacy and security goals, and introduce too much risk for our customers. We have brought it up internally multiple times, due to this discussion and other ways customers has requested this feature. But we have been very hesitant every time, and then it was forgotten again. We feel like it's more fair that we close this and explain why instead of just let it sit open for a few more years.

Is it technically hard to implement? Yes, it is not trivial. There is no single switch to flip in the code to make the current split tunneling feature work backwards. It's way more intricate than that. The entire app is written ground up with the design goal to securely absorb all network traffic. Implementing inverse split tunneling is by no means isolated to the code dealing with the current split tunneling feature. There are many layers of safety mechanisms incorporated to make sure the user's traffic does not leak. Making the app not capture traffic by default, and only capture some traffic, would invert this entire fundamental design. This would include large changes to how we manage the routing table, firewall rules, DNS settings and more. If we were to implement this we would risk introducing bugs that would put users that do want to capture all traffic at risk. If there are legitimate code paths where the users' traffic should go outside the tunnel, it is much easier to accidentally make users leak that should not leak, compared to if these code paths would not even exist.

Even if we were to implement inverse split tunneling without any bugs, we think this is a huge foot-gun for our customers. If there is a way to configure the app to leak all traffic by default, we suspect some users might unintentionally get into that state. And no, I do not mean that we would implement the feature with terrible UX, nor that we think our users are stupid. This could easily happen if for example a browser upgrade changes the browser footprint in such a way that the code identifying which apps should go in the tunnel fails to match the browser after the upgrade. Now the browser would suddenly leak outside the tunnel, when the user intended for it to go in the tunnel. It might take a very long time before the user notice this leak. This is simply not acceptable. Yes, this problem could already happen with our current split tunneling. But the important difference is that if this happens now, an excluded browser would suddenly communicate in the tunnel instead of outside it. This is most of the time a way less dangerous change than the opposite. Always fail closed, never fail open.

Our recommendation would be that users that only wish to tunnel certain traffic should instead use the vanilla WireGuard app, and set it up so that it only captures certain traffic. It's not perfect, and it misses out on a lot of the other features that our app has sadly, but it would be the only way if you really need to capture only a small subset of all traffic on your device.

We hope you understand our motivation. We want to focus on privacy and security. Thank you.

**Comment by devingDev** [comment]
If you wanna assume your customers are dumb and don't know what they're doing great!
I already switched so not anymore interested anyway
Have a great day

**Comment by JaxxArmstrong** [comment]
There we have it black on white. They won't implement this to many crucial feature that would simplify a lot of user's daily life and we are now forced to either fork the project and implement it ourselves or are delegated to use non-preferred VPN products from other vendors.

Note that this is NOT a technical implementation, but they don't want to rewrite part of their app. Why this is is up to everyone to ponder on, but laziness comes to mind. I understand they aren't making bank on their product, but this is totally unacceptable.

Another customer switching to a competing product.

**Comment by GRIFFiin** [comment]
Maybe you could implement it as a new branch?? Like only "expert" would know how branch works in github, and we could bug report anything

**Comment by faern** [comment]
I'm locking this conversation so that the [official response](https://github.com/mullvad/mullvadvpn-app/discussions/9050#discussioncomment-14878259) does not get lost in the thread. I imagine a lot of customers who find this thread are interested in Mullvad's stance in this. I do not think further posts will be constructive anyway.
